carp.common/src/commonMain/kotlin/dk/cachet/carp/common/infrastructure/services/ApplicationServiceInvoker.kt

```diff
@@ -0,0 +1,16 @@
+package dk.cachet.carp.common.infrastructure.services
+
+import dk.cachet.carp.common.application.services.ApplicationService
+
+
+/**
+ * Support invoking [TRequest] on a specified [TService].
+ */
+interface ApplicationServiceInvoker<
+    TService : ApplicationService<TService, *>,
+    TRequest : ApplicationServiceRequest<TService, *>
+>
+{
+    suspend fun TRequest.invoke( service: TService ): Any?
```

Comment:
Instead, it would be possible to _not_ make `TRequest` part of the `ApplicationServiceInvoker` type definition. But, then implementing classes become unwieldly. They'd all have to cast from `ApplicationServiceRequest` to the concrete requests. This basically centralizes the cast from `Any?` to `TReturn` in a single place.

Reply:
> No, since `ApplicationServiceInvoker` is an abstraction for _all_ requests on an application service. It can't include a generic `TReturn` (it's star-projected right now) since there is no "one type" (other than `Any?`, defeating the purpose) which would work for all requests in the service.

I mean, I kind of get what you say, it's just that my alarm rings anytime I see `Any` written anywhere (maybe a habit that stuck with me after developing in Typescript). My other suggestion would have been to have an interface of some sorts that all return types must implement. But I also feel like that would be a stretch since there aren't any commonalities between any of the return types so it can't be anything else but an empty interface. But also then if you wanted to have primitive or built-in types into a library as return types than you would have to wrap them and that is a bit annoying.

Reply:
> Instead, it would be possible to _not_ make `TRequest` part of the `ApplicationServiceInvoker` type definition. But, then implementing classes become unwieldly. They'd all have to cast from `ApplicationServiceRequest` to the concrete requests. This basically centralizes the cast from `Any?` to `TReturn` in a single place; the decorator.

No I don't think there is a need for that.

Reply:
I looked a bit more into what was possible, and gave up. In fact, I changed `ICommand<TRequest>` to also just return `Any?` since having it return `TReturn` just forces people using that interface to do unchecked `TReturn` casts. It doesn't add value.

I did figure out that it makes more sense for `ApplicationServiceDecorator.invoke` to take `ApplicationServiceRequest<TService, TReturn>` rather than the more specific `TRequest`. It has the benefit type inference is enabled for `TReturn`, based on the passed specific request object, as opposed to the caller having to know the return type. As you can see in the last commit, it simplifies creating decorators for application services, and, it provides more typesafety for callers.

I think I'll leave the generics juggling for the decorators at that. 😅